MiqExpression::Target.parse returns null object to simplify callers

kbrock opened the pull request:
redo of https://github.com/ManageIQ/manageiq/pull/22451
Main difference being that an `MiqExpression::Invalid` object is now used.

I had rebased that PR and it got away from me.

The goal is to drop get_col_info, get all column metadata operations out of `MiqExpression` and into `MiqExpression::Target`.

This is 11 commits. lots of 📖 commits

- Have all `Target` subclasses respond to the same methods
- remove special cases when dealing with `#parse` returned objects.
- remove unchecked `ArgumentError`, especially when we don't need to handle them.
- Use a `Symbol`/`nil` for `column_type`. Before it was typically a `String` or `Symbol`, but sometimes a `nil`)

kbrock commented:
@Fryguy this is a bit of a distraction.
Help me figure out how to cut this down to a smaller chunk.

I really like the concept of returning an `Invalid` object. gets us closer to dropping `get_col_info`.

Going from here, if we get the field into the MiqExpression tree, I think that buys us a lot across the board, and many of these `column_type` calls will go away (since we'll have the actual `MiqExpression::Field`. So I got stuck on the whole "do I change this to clean it up knowing it will go away or do I try and make it go away and not make a big PR like this"

So I punted but you saw that. If you have a suggestion on which commits in this PR are easy to include, then I can rebase / split and we can get it in without too much though

Fryguy commented on lib/miq_expression/invalid.rb:
```diff
@@ -0,0 +1,21 @@
+class MiqExpression::Invalid < MiqExpression::Target
```
`MiqExpression::Invalid` makes it sound like an invalid _expression_ as opposed to an invalid _target_.  This might be clearer:

```suggestion
class MiqExpression::InvalidTarget < MiqExpression::Target
```

kbrock commented:
update:

- fixed extra whitespace
- renamed Invalid to InvalidTarget
- fixed unused parameter rubocop (add underscore to options)
-

Fryguy commented on lib/miq_expression.rb:
```diff
@@ -1370,10 +1365,10 @@ def fields(expression = exp)
-        tg = self.class.parse_field_or_tag(expression["value"].to_s)
-        ret << tg if tg
+        tg = Target.parse(val)
+        ret << tg unless tg.kind_of?(InvalidTarget)
```
Should this be something like

```suggestion
        ret << tg if tg.valid?
```

kbrock replied:
Discussed this w/ Jason

I had originally intended on making this `tg.valid?`, but `.valid?` is a little more stringent.

And the future play is to drop this parsing, so I didn't want to have future code question if we need to add a `valid?` check.

Fryguy commented on lib/miq_expression/tag.rb:
```diff
@@ -58,6 +51,10 @@ def arel_attribute
     target && target.arel_table[:id, arel_table]
   end
 
+  def valid?
+    !!target
+  end
```
What is the difference between valid? returning false and an InvalidTarget?

kbrock replied:
the syntax for a tag is valid. But the reference in value is invalid.
(syntax ok, semantics bad)

Fryguy commented on lib/miq_expression.rb:
```diff
@@ -726,48 +717,50 @@ def self.operands2rubyvalue(operator, ops, context_type)
-    case typ.to_s
-    when "string", "text", "boolean", nil
+    case typ&.to_sym
+    when nil
+      "\"string\""
+    when :string, :text, :boolean
```
This looks wrong - with nil it should return the `value` not the literal word "string".  So I think this should be

```suggestion
    when :string, :text, :boolean, nil
```

Fryguy commented on lib/miq_expression.rb:
```diff
@@ -816,7 +809,9 @@ def self.quote_human(val, typ)
         val
       end
-    when "string", "date", "datetime"
+    when nil
+      "\"string\""
+    when :string, :date, :datetime
```
Similar comment to https://github.com/ManageIQ/manageiq/pull/22519/files#r1207226074

kbrock commented:
update:
- added specs to test quote and quote_human passing in a `string`, `integer`, and `nil`.
- fixed `quote()` for `nil` case.
- dropped `|| :string` in a few cases that were passing directly into `quote()`

kbrock commented on app/models/condition.rb:
```diff
@@ -120,12 +120,12 @@ def self._subst(rec, opts, tag, mode)
         value = ref.nil? ? "" : Tag.list(ref, :ns => tag)
       end
-      value = MiqExpression.quote(value, ohash[:type] || "string")
+      value = MiqExpression.quote(value, ohash[:type]&.to_sym)
```
this has been changed to drop the default to `string` (2 places).

But since string and nil should do the same thing in `quote()` and there are other places in the code that call `quote()` without defaulting to string, I decided to drop

kbrock commented:
codacity is hung. it says it is 🍏 over there

Fryguy commented:
Backported to `petrosian` in commit 757e76d3a328449b933346b46b3f93317d664526.

```text
commit 757e76d3a328449b933346b46b3f93317d664526
Author: Jason Frey <fryguy9@gmail.com>
Date:   Tue May 30 16:01:25 2023 -0400

    Merge pull request #22519 from kbrock/miq_expression_parse
    
    MiqExpression::Target.parse returns null object to simplify callers
    (cherry picked from commit 5fb8ce701dcdcfb071daa901d63b419dc73fc0c2)
```
